nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/io/avro/TestAvroSchemaFactoryService.java

```diff
@@ -0,0 +1,228 @@
+import org.nuxeo.runtime.test.runner.Features;
+import org.nuxeo.runtime.test.runner.FeaturesRunner;
+import org.nuxeo.runtime.test.runner.RuntimeFeature;
+
+/**
+ * @since 10.2
+ */
+@RunWith(FeaturesRunner.class)
+@Features(RuntimeFeature.class)
+@Deploy("org.nuxeo.ecm.core")
+@Deploy("org.nuxeo.ecm.core.io")
+@Deploy("org.nuxeo.ecm.core.schema")
+@Deploy("org.nuxeo.runtime.stream")
+public class TestAvroSchemaFactoryService {
+
+    private static Collection<String> FORBIDDEN = Arrays.asList("-", "__tutu__", "-gru__du__buk-");
+
+    private static final String FIELD = "field";
+
+    protected static final Map<String, String> TYPES_MAPPING = new HashMap<>();
+
+    static {
+        TYPES_MAPPING.put("date", "timestamp-millis");
+        TYPES_MAPPING.put("binary", "bytes");
+    }
+
+    @Inject
+    public AvroSchemaFactoryService service;
+
+    @Inject
+    public SchemaManager schemaManager;
+
+    @Test
+    @Deploy("org.nuxeo.ecm.core.test.tests:OSGI-INF/test-repo-core-types-contrib.xml")
+    public void testComplexType() {
+        DocumentType type = schemaManager.getDocumentType("ComplexDoc");
+        AvroSchemaFactoryContext context = service.createContext();
+        // schema creation
+        Schema avro = context.createSchema(type);
+        assertNotNull(avro);
+        // 3 root schemas
+        Field common = avro.getField("common");
+        assertNotNull(common);
+        Field dublincore = avro.getField("dublincore");
+        assertNotNull(dublincore);
+        Field complexschema = avro.getField("complexschema");
+        assertNotNull(complexschema);
+        // exploration of complexschema
+        assertEquals("cmpf", complexschema.schema().getNamespace());
+        Field attachedFile = complexschema.schema().getField("attachedFile");
+        Schema fileext = attachedFile.schema();
+        assertEquals("fileext", fileext.getName());
+        assertSame(STRING, fileext.getField("name").schema().getType());
+        Schema vignettes = fileext.getField("vignettes").schema();
+        assertEquals("array", vignettes.getType().getName());
+        Schema vignette = vignettes.getElementType();
+        assertSame(STRING, vignette.getField("label").schema().getType());
+        assertSame(LONG, vignette.getField("width").schema().getType());
+        assertSame(LONG, vignette.getField("height").schema().getType());
+        Schema content = vignette.getField("content").schema();
+        assertNull(content.getField("mime-type"));
+        assertSame(STRING, content.getField("name").schema().getType());
+        assertSame(STRING, content.getField("digest").schema().getType());
+        assertSame(STRING, content.getField("encoding").schema().getType());
+        assertSame(STRING, content.getField(context.replaceForbidden("mime-type")).schema().getType());
+        assertSame(LONG, content.getField("length").schema().getType());
+        assertSame(BYTES, content.getField("data").schema().getType());
```

Comment:
Could we also have a test that shows the full JSON serialized format for a Nuxeo type? This will provide non-regression tests.